### PR TITLE
Demo on Convolutions that includes:

### PR DESCRIPTION
**Pull request by jpilaul**
- SQL expressions
- Jseval
- Explanation of convolutions

### CODE REVIEW

**Comment by mailletf** [comment]
We should consider trying to do it with tensorflow also if it's easy.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 1]
```diff
@@ -0,0 +1,386 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Image Processing with Convolutions\n",
+    "\n",
+    "In image processing, most image filters and image transformation use convolutions. Convolutions modify the original matrix of pixels through a pointwise multiplication with a kernel or filter matrix. Wikipedia describes <a href=\"https://en.wikipedia.org/wiki/Kernel_(image_processing)\">convolutions on images</a> as:\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: and image transformation -> and image transformationS

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    "\n",
+    "> Convolution is the process of multiplying each element of the image with its local neighbors, weighted by the kernel. For example, if we have two three-by-three matrices, one a kernel, and the other an image piece, convolution is the process of flipping both the rows and columns of the kernel and then multiplying locationally similar entries and summing. The [2,2] element of the resulting image would be a weighted combination of all the entries of the image matrix, with weights given by the kernel: \n",
+    "\n",
+    "> ![Image of convolution](https://wikimedia.org/api/rest_v1/media/math/render/svg/1a5bdd585d515770c888ea5b4ea07a7a5166cc8d)\n",
+    "\n",
+    "Amongst the suite of applications of convolutions, image blurring and sharpening as well as edge detection are the most common. In this demo, we will use MLDB query to transform efficiently transform images. To do so, we will use the [MNIST database of handwriten digits](http://yann.lecun.com/exdb/mnist/).\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: reformuler: we will use MLDB query to transform efficiently transform images

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    "\n",
+    "In this demo, we will use a custom function type called [SQL Expression Function](../../../../doc/#builtin/functions/SqlExpressionFunction.md.html) and [jseval](../../../../doc/#builtin/sql/ValueExpression.md.html) for inline definition of functions using Javascript."
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: le lien à jseval devrait avoir le anchor #jseval

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    "from ipywidgets import widgets, interact"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Loading the data\n",
+    "\n",
+    "A pickled version of the dataset is available on the deeplearning.net website http://deeplearning.net/tutorial/gettingstarted.html.\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: met le lien sur le texte "on the deeplearning.net website"

File: container_files/demos/Convolutions using SQL and Javascript.ipynb (hunk 3 above)
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: ?-> In this demo, we will use the jseval function to execute Javascript code inline with SQL, and the SQL Expression Function to persist and be able to easily reuse that piece of Javascript.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    "\n",
+    "The dataset has been unpickled and saved in a public Amazon's S3 cloud storage. Check out MLDB's [Protocol Handlers](../../../../doc/#builtin/Url.md.html) for Files and URLS."
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: and URLS for more details on loading remote ressources.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    "})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Taking a random image and starting image manipulation\n",
+    "\n",
+    "Similarly to first few steps in the [Real-Time Digits Recognizer](../../../../ipy/notebooks/_tutorials/v2016.06.22.dev-5d85bf1/Real-Time%20Digits%20Recognizer.ipynb) demo, we will display random MNIST digits from the test set."
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: to THE first

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 331,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from pymldb import Connection\n",
+    "mldb = Connection(host = \"http://127.0.0.1:11700/\")"
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: simplement: mldb = Connection()

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+      "text/plain": [
+       "<matplotlib.figure.Figure at 0x7f5fb0240190>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: tu peux faire ceci avec le dataset sampled:

```
data = mldb.query("""
    SELECT * EXCLUDING(label) FROM sample(digits_mnist, {rows: 1})
""")

image = data.as_matrix().reshape(28, 28)
plt.imshow(image)
plt.gray()
```

Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: Si tu veux du test set uniquement tu peux le faire aussi mais c'est plus lent à rouler un peu:

```
 SELECT * EXCLUDING(label) 
FROM sample(
                (select * from digits_mnist where rowHash() % 5 = 0),
                {rows: 1})
```

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    "create_convolution()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As seen in the code above, each new pixel in the convolved picture is the weighted sum of the the pixel and its neighboring pixels where the weights are the values in the kernel matrix.  \n",
+    "\n",
+    "Doing convolutions with custom function type called [SQL Expression Function](../../../../doc/#builtin/functions/SqlExpressionFunction.md.html) and [jseval](../../../../doc/#builtin/sql/ValueExpression.md.html) for inline definition of functions using Javascript allows us to process large amounts of data using the optimizations inherent to the MLDB platform. Convolutions are typically very time consuming operations with \n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: function OF type

Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: to the MLDB platform -> to MLDB

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    "$O(n.r^2)$ complexity in this case where n is the number of features and r is the radius (i.e. neighboring pixels).\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: n \dot r^2  ?

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    "\n",
+    "There were two steps to creating the function above:\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: je bougerais ça avant le bloc de code. below there will be these two things. ça va orienter le lecteur

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "options=('Right Sobel', 'Detect Edges', 'Sharpen', 'Box Blur', 'Approximated Gaussian Blur')\n",
+    "\n",
+    "# Creating an interactive dropdown menu\n",
+    "interact(convolutionFunc, image_processing=options, );\n",
+    "print \"Choose an image processing option from the drop-down menu\""
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: peut-être un gif animé pour teaser le monde

Comment (jpilaul) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: je n'ai pas trouvé comment faire ça...

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+    "        \"type\": \"sql.expression\",\n",
+    "        \"params\": {\n",
+    "            \"expression\": JsConvolutionExpr,\n",
+    "            \"prepared\": True\n",
+    "        }\n",
+    "    })\n",
+    "    \n",
+    "create_convolution()"
+   ]
+  },
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: Il manque un paragraphe qui explique ce qui se passe vite vite. Genre ici on va prendre l'image d'en haut et appliquer le filtre choisi dans le dropdown balb la. Et au minimum mettre une note en bold qui dit que ceci est interatif et d'utiliser le live netbook pour l'essayer.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+    "I found the 'Detect Edges' convolution particularly useful when training image recognition models. This can be useful in many Machine Vision applications."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Convolutions with MLDB's TensorFlow plug-in\n",
+    "\n",
+    "Not everyone will want to code their own convolutions from scratch (such as with the `create_convolution()` function above). In fact, given the myriad of tools available, it may save you time and effort to use external librairies. MLDB has integrated the TensorFlow Open Source Library for Machine Intelligence allowing us to leverage some of the great Computer Vision APIs that it offers. Let's get started with the same images as before.\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: that it offers, as well as GPU acceleration.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+    "\n",
+    "First, I reshape my image and kernel lists into 4D tensors. Then, I use the [`tf_Conv2D`](../../../../doc/builtin/sql/ValueExpression.md.html#builtinfunctions) built-in convolution function."
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: use the [`tf_Conv2D`], which is a Tensorflow operator that is exposed as an MLDB built-in function directly in SQL.

Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: into 4D tensor in the NHWC tensor format.

File: container_files/demos/Convolutions using SQL and Javascript.ipynb [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+    "interact(TensorFlowConvolution, image_processing=kernelDict.keys(), );\n",
+    "print \"Choose an image processing option from the drop-down menu\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here are a few definitions:\n",
+    "* Batch: The data is sometimes split into batches to parallelize Neural Network training and provide more frequent updates Neural Network weights. For more information, see Carnegie Mellon University's paper on [Efficient Mini-batch Training for Stochastic Optimization](https://www.cs.cmu.edu/~muli/file/minibatch_sgd.pdf). In our example, we only have one unsplit batch.\n",
```
Comment (mailletf) [inline container_files/demos/Convolutions using SQL and Javascript.ipynb]: ici en fait ce n'est pas pour des mini-batch. c'est pour paralleliser le processing sur le GPU. si tu as plusieurs convolutions à faire tu peux le faire avec 1 call et ça va rouler plus vite que n call.

**Comment by mailletf** [comment]
faudrait renomer le fichier pour qu'il match le titre du notebook

**Comment by mailletf** [comment]
+1
